tmt/steps/prepare/brand_new_allmighty_install/providers/__init__.py

```diff
@@ -0,0 +1,147 @@
+        """
+        Parse and validate the artifact identifier.
+
+        :param artifact_id: The raw artifact identifier
+        :return: The parsed identifier specific to this provider
+        :raises ValueError: If the artifact ID format is invalid
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def list_artifacts(self) -> Iterator[ArtifactInfoT]:
+        """
+        List all available artifacts for a given build
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def _download_artifact(self, artifact: ArtifactInfoT, guest: Guest, destination: Path) -> None:
+        """
+        Action: Download a single artifact to the specified destination.
+
+        :param guest: The guest where the artifact should be downloaded
+        :param destination: Path where the artifact should be downloaded
+        """
+        raise NotImplementedError
+
+    def download_artifacts(
+        self,
+        guest: Guest,
+        download_path: Path,
+        exclude_patterns: list[Pattern[str]],
+    ) -> list[Path]:
+        """
+        Download all artifacts to the specified destination.
+
+        :param guest: The guest where the artifacts should be downloaded.
+        :param download_path: The local path to save the downloaded artifact.
+        :param exclude_patterns: Patterns to exclude certain files from being downloaded.
+        :returns: A list of paths to the downloaded artifacts.
+        :raises GeneralError: Unexpected errors outside the download process.
+        :note: Errors during individual artifact downloads are
+            caught, logged as warnings, and ignored.
+        """
+        self.logger.info(f"Downloading artifacts to {download_path!s}")
+
+        # Ensure download directory exists on guest (create only if missing)
+        guest.execute(
+            ShellScript(
+                f"[ -d {quote(str(download_path))} ] || "
+                f'{"sudo " if not guest.facts.is_superuser else ""}'
+                f"mkdir -p {quote(str(download_path))}"
+            ),
+            silent=True,
+        )
+
+        downloaded_paths: list[Path] = []
+        artifact_count = 0
+
+        for artifact in self._filter_artifacts(exclude_patterns):
+            local_path = download_path / str(artifact)
+            self.logger.debug(f"Downloading {artifact} to {local_path}")
+
+            try:
+                self._download_artifact(artifact, guest, local_path)
+                downloaded_paths.append(local_path)
+                artifact_count += 1
+                self.logger.info(f"Downloaded {artifact} to {local_path}")
+
+            except DownloadError as err:
+                # Warn about the failed download and move on
+                self.logger.warning(f"Failed to download {artifact}: {err}")
+
+            except Exception as err:
+                raise GeneralError(f"Unexpected error downloading {artifact}") from err
```

Comment:
A general note on error messages and exceptions: tmt uses slightly different form of including interesting bits in the message, it's usually wrapped with single-quotes, rarely using `:`, and there should be a dot at the end. See https://tmt.readthedocs.io/en/stable/contribute.html#introduction

```python
f"Unexpected error downloading '{artifact}'."
f"Successfully downloaded {artifact_count} artifacts."
"Failed to initialize API session."
f"API call '{method}' failed."
f"Invalid artifact ID format '{artifact_id}'."
```

And so on. Please, review the raised exceptions and logging sites.